pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go

```diff
@@ -22,319 +22,344 @@ import (
-	).register()
+		"index", "child",
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type(
+					(*scpb.PrimaryIndex)(nil),
+					(*scpb.SecondaryIndex)(nil),
+				),
+				to.Type(
+					(*scpb.IndexName)(nil),
+					(*scpb.IndexPartitioning)(nil),
+					(*scpb.SecondaryIndexPartial)(nil),
+					(*scpb.IndexComment)(nil),
+				),
+				joinOnIndexID(from, to),
+				targetStatusEq(fromTarget, toTarget, scpb.ToAbsent),
+				currentStatus(fromNode, scpb.Status_VALIDATED),
+				currentStatus(toNode, scpb.Status_ABSENT),
+			}
+		},
+	)
 
-	depRule(
+	registerDepRule(
 		"dependents removed before index",
 		scgraph.Precedence,
-		scpb.ToAbsent,
-		element(scpb.Status_ABSENT,
-			(*scpb.IndexName)(nil),
-			(*scpb.IndexPartitioning)(nil),
-			(*scpb.SecondaryIndexPartial)(nil),
-			(*scpb.IndexComment)(nil),
-		),
-		element(scpb.Status_ABSENT,
-			(*scpb.PrimaryIndex)(nil),
-			(*scpb.SecondaryIndex)(nil),
-		),
-		screl.DescID,
-		screl.IndexID,
-	).register()
+		"dependent", "index",
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type(
+					(*scpb.IndexName)(nil),
+					(*scpb.IndexPartitioning)(nil),
+					(*scpb.SecondaryIndexPartial)(nil),
+					(*scpb.IndexComment)(nil),
+				),
+				to.Type(
+					(*scpb.PrimaryIndex)(nil),
+					(*scpb.SecondaryIndex)(nil),
+				),
+				joinOnIndexID(from, to),
+				targetStatusEq(fromTarget, toTarget, scpb.ToAbsent),
+				currentStatusEq(fromNode, toNode, scpb.Status_ABSENT),
+			}
+		},
+	)
 }
 
 // These rules ensure that before an offline-backfilled index can begin
 // backfilling, the corresponding temporary index exists in WRITE_ONLY.
 func init() {
-	var (
-		from, fromTarget, fromNode         = targetNodeVars("from")
-		to, toTarget, toNode               = targetNodeVars("to")
-		descID, tempIndexID        rel.Var = "desc-id", "temp-index-id"
-	)
 	registerDepRule(
 		"temp index is WRITE_ONLY before backfill",
 		scgraph.Precedence,
-		fromNode, toNode,
-		screl.MustQuery(
-			from.Type(
-				(*scpb.TemporaryIndex)(nil),
-			),
-			to.Type(
-				(*scpb.PrimaryIndex)(nil),
-				(*scpb.SecondaryIndex)(nil),
-			),
-			descID.Entities(screl.DescID, from, to),
-			to.AttrEqVar(screl.TemporaryIndexID, tempIndexID),
-			from.AttrEqVar(screl.IndexID, tempIndexID),
-
-			fromTarget.AttrEq(screl.TargetStatus, scpb.Transient.Status()),
-			toTarget.AttrEq(screl.TargetStatus, scpb.ToPublic.Status()),
-
-			fromNode.AttrEq(screl.CurrentStatus, scpb.Status_WRITE_ONLY),
-			toNode.AttrEq(screl.CurrentStatus, scpb.Status_BACKFILLED),
-
-			screl.JoinTargetNode(from, fromTarget, fromNode),
-			screl.JoinTargetNode(to, toTarget, toNode),
-		),
+		"temp", "index",
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type((*scpb.TemporaryIndex)(nil)),
+				to.Type((*scpb.PrimaryIndex)(nil), (*scpb.SecondaryIndex)(nil)),
```

Comment:
I wish there was a way to make these type unions pop out more systematically somehow. It's why I originally newlined each subtype, I found that it helped readability. I think this is fine though.

Reply:
Done.